### PR TITLE
feat(agent): add prometheus metrics exposition

### PR DESCRIPTION
**Pull request by aurrelhebert**

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -82,11 +136,14 @@ pub fn run_core(cfg: Config) {
...
                 Ok(msg) => Some(msg),
                 Err(kafka_error) => {
+                    KAFKA_ERROR_COUNTER.add(1);
                     error!("Error while receiving from Kafka: {:?}", kafka_error);
                     None
                 }
             }
         }).for_each(|msg| {
+
+            PROCESS_COUNTER.add(1);
```
Comment (FlorentinDUBOIS) [inline src/main.rs]: PROCESS_COUNTER.inc();

File: src/main.rs [hunk 2]
```diff
@@ -113,11 +170,13 @@ pub fn run_core(cfg: Config) {
                     .map_err(|e| e.description().to_string())
             })
             .or_else(|err| {
+                ERROR_COUNTER.add(1);
```
Comment (FlorentinDUBOIS) [inline src/main.rs]: ERROR_COUNTER.inc();

File: src/main.rs [hunk 3]
```diff
@@ -113,11 +170,13 @@ pub fn run_core(cfg: Config) {
...
                 error!("Error while processing message: {:?}", err);
                 Ok(())
             });
 
             handle.spawn(process_message).unwrap();
+            SUCCESS_COUNTER.add(1);
```
Comment (FlorentinDUBOIS) [inline src/main.rs]: SUCCESS_COUNTER.inc();

File: src/main.rs [hunk 4]
```diff
@@ -33,20 +37,70 @@ use futures_cpupool::Builder;
...
+    let metrics_server = path!("metrics")
+        .map(|| {
+            let encoder = TextEncoder::new();
+            let metric_families = prometheus::gather();
+            let mut buffer = vec![];
+            encoder.encode(&metric_families, &mut buffer).unwrap();
+            buffer
+        });
+
+    serve(metrics_server)
```
Comment (FlorentinDUBOIS) [inline src/main.rs]: You have to run this in another thread:

```rust
use std::thread;
...
fn main() {
...
    let metrics_thread = thread::spawn(move || { serve(metrics_server).run(([127, 0, 0, 1], 3030)); });

    run_core(cfg);

    metrics_thread.join().expect("stop metrics server thread without troubles");
}
```
https://doc.rust-lang.org/std/thread/fn.spawn.html

File: src/main.rs [hunk 5]
```diff
@@ -33,20 +37,70 @@ use futures_cpupool::Builder;
...
+use warpservice::{warp10_post, DnsQueryResults};
 use warp10::Label;
 
+lazy_static! {
+    static ref KAFKA_ERROR_COUNTER: IntGauge = register_int_gauge!(opts!(
+        "dns_agent_kafka_error_events",
+        "Number of events causing kafka consumer error",
+        labels! {"handler" => "all",}
+    ))
+    .unwrap();
```
Comment (FlorentinDUBOIS) [inline src/main.rs]: replace `unwrap` by `expect("create dns_agent_kafka_error_events metrics")`

File: src/main.rs [hunk 6]
```diff
@@ -33,20 +37,70 @@ use futures_cpupool::Builder;
...
+
+    static ref PROCESS_COUNTER: IntGauge = register_int_gauge!(opts!(
+        "dns_agent_processed_events",
+        "Number of events processed",
+        labels! {"handler" => "all",}
+    ))
+    .unwrap();
```
Comment (FlorentinDUBOIS) [inline src/main.rs]: replace `unwrap` by `expect("create dns_agent_processed_events metrics")`

File: src/main.rs [hunk 7]
```diff
@@ -33,20 +37,70 @@ use futures_cpupool::Builder;
...
+    
+    static ref ERROR_COUNTER: IntGauge = register_int_gauge!(opts!(
+        "dns_agent_error_events",
+        "Number of events causing error when processed",
+        labels! {"handler" => "all",}
+    ))
+    .unwrap();
```
Comment (FlorentinDUBOIS) [inline src/main.rs]: replace `unwrap` by `expect("create dns_agent_error_events metrics")`

File: src/main.rs [hunk 8]
```diff
@@ -33,20 +37,70 @@ use futures_cpupool::Builder;
...
+
+     static ref SUCCESS_COUNTER: IntGauge = register_int_gauge!(opts!(
+        "dns_agent_success_events",
+        "Number of events correctly handled",
+        labels! {"handler" => "all",}
+    ))
+    .unwrap();
```
Comment (FlorentinDUBOIS) [inline src/main.rs]: replace `unwrap` by `expect("create dns_agent_success_events metrics")`

**Review by FlorentinDUBOIS — CHANGES_REQUESTED** [review]

File: src/main.rs (hunk 1 above)
Comment (aurrelhebert) [reply src/main.rs]: Deny :) It's a gauge

File: src/main.rs (hunk 2 above)
Comment (aurrelhebert) [reply src/main.rs]: Deny :) It's a gauge

File: src/main.rs (hunk 3 above)
Comment (aurrelhebert) [reply src/main.rs]: Deny :) It's a gauge

File: src/main.rs (hunk 5 above)
Comment (aurrelhebert) [reply src/main.rs]: https://docs.rs/prometheus/0.5.0/prometheus/#types ?

File: src/main.rs (hunk 6 above)
Comment (aurrelhebert) [reply src/main.rs]: https://docs.rs/prometheus/0.5.0/prometheus/#types ?

File: src/main.rs (hunk 7 above)
Comment (aurrelhebert) [reply src/main.rs]: https://docs.rs/prometheus/0.5.0/prometheus/#types ?

File: src/main.rs (hunk 8 above)
Comment (aurrelhebert) [reply src/main.rs]: https://docs.rs/prometheus/0.5.0/prometheus/#types ?

**Review by FlorentinDUBOIS — APPROVED** [review]

**Pull request merged**